services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java

```diff
@@ -233,6 +233,47 @@ private JsonObject getCredentials(final String tenantId, final String authId, fi
         return null;
     }
 
+    @Override
+    public void addCredentials(String tenantId, JsonObject otherKeys, Handler<AsyncResult<CredentialsResult<JsonObject>>> resultHandler) {
+        CredentialsResult credentialsResult = addCredentialsResult(tenantId, otherKeys);
+        resultHandler.handle(Future.succeededFuture(credentialsResult));
+    }
+
+    private CredentialsResult addCredentialsResult(final String tenantId, JsonObject otherKeys) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(otherKeys);
+
+        String deviceId = otherKeys.getString(FIELD_DEVICE_ID);
+        String authId = otherKeys.getString(FIELD_AUTH_ID);
+
+        Map<String, JsonArray> credentialsForTenant = credentials.get(tenantId);
+
+        if (credentialsForTenant == null) {
+            credentialsForTenant = new HashMap<>();
+            credentials.put(tenantId, credentialsForTenant);
+        }
+
+        JsonArray authIdCredentials = credentialsForTenant.get(authId);
+        if (authIdCredentials == null) {
+            authIdCredentials = new JsonArray();
+        } else {
+
+            // check if already exist credentials with the type and auth-id for the device-id from the payload.
+            for (Object credentialsObj: authIdCredentials) {
+                JsonObject credentials = (JsonObject) credentialsObj;
+                if (credentials.getString(FIELD_TYPE).equals(otherKeys.getString(FIELD_TYPE)) &&
```

Comment:
Can otherKeys.getString(FIELD_TYPE) be null?
And if yes: could that corrupt the file content?
Did not have time to fully check it, so just take it as a thought from my side to have a closer look about if there could be checks missing that would prevent corrupting the file content by any means...

Reply:
The "type" is mandatory. In any case the user could make a mistake providing a not valid JSON. I'll consider it (opening an issue after this first implementation will be merged).

Reply:
That is a good plan :-)